### DISCUSSION TITLE
x-on should not execute the execution result of the code.

### DISCUSSION THREAD

**Discussion by JohnTing** [alpinejs/alpine#4047] [7. Bugs (archived)]
in alpinejs : 3.13.5
https://alpinejs.dev/directives/on#the-event-object
```l!
In addition, Alpine also passes the event object to any methods referenced without trailing parenthesis.
```

It should not execute the execution result of the code.

For example: if I want to store the `foo1` methods into `storedFoo1` when the button is clicked.
```javascript
function foo1() {
    console.log('foo1');
}
```

In JavaScript, 'foo1' is not executed upon clicking; it is simply stored.
```html
// var storedFoo1=null;
<button onclick="storedFoo1=foo1">vanilla js</button>
```
In Vue3, 'foo1' is not executed upon clicking; it is simply stored.
```html
// const storedFoo1 = ref(null);
<button @click="storedFoo1=foo1">vuejs</button>
```
In Alpine, 'foo1' is executed upon clicking; 
```html
<div x-data="{ storedFoo1: null }" >
    <button x-on:click="storedFoo1=foo1">alpinejs</button>
</div>
        
```

---

Another example:
The delayedMessage method has a return method used for canceling events.

```javascript
function delaydMessage(message, millisecond) {
    const id = setTimeout(()=> {console.log(message)}, millisecond);
    // used to cancel
    return () => {clearTimeout(id);console.log('cancel!');};
}
```

In Alpine, it will be canceled immediately.
```html
<button x-on:click="delaydMessage('hello world', 3000)">click</button>
```

---

A way to avoid the behavior is to add something else to the first script.

like this
```html
<button x-on:click="null;delaydMessage('hello world', 3000)">click</button>
```
or
```html
<button x-on:click="'';delaydMessage('hello world', 3000)">click</button>
```
And this one will throw error.
```html
<button x-on:click=";delaydMessage('hello world', 3000)">click</button>
```

**Comment by SimoTod** [comment]
I'll start with the second example. So it's a design choice made by the maintainer. 

In Alpine, if the directive content resolves to a callable, it gets invoked. This is to allow people to use the function name without parenthesis in you x-on:click and that will just work as described here: https://alpinejs.dev/directives/on#the-event-object

This is done by the evaluator which eventually calls https://github.com/alpinejs/alpine/blob/main/packages/alpinejs/src/evaluator.js#L137

The downside is that, if you are already calling a function and that returns a callable, the returned value gets invoked to. To avoid it, you can also do
```html
<button x-on:click="() => delaydMessage('hello world', 3000)">click</button>
```
I don't think it can be changed as it's too much of a change for the way Alpine works. Maybe it should be better documented (the documentation is open source if you want to send a PR).

The first example is admittedly a bad side effect of the above, the result of the assignment is returned (your function reference) by the evaluation and for the same reason, being a function, it gets invoked. not sure there's going to be a smart way to prevent this in the apline source code.

**Comment by ekwoka** [comment]
Would definitely be a breaking change to have it not call it.

I do mostly struggle to see use cases that would cause issue with this. I've made some quite wacky and wild components, and never had this one crop up. Like why does `delayedMessage` return a cancel callback if you aren't going to cancel it? and why is the reassigning methods on the context?

Not to say they couldn't ever happen, it just feels like a strange XY problem that might have a more sensible solution than this approach to begin with.

But, anyway, it is an intentional decision as mentioned.

Another "add more" you could do is just `delayedMessage(args) && true` or similar.

maybe there could be an argument for a more global directive modifier for `.nocall` that will not call returned functions, which would be relatively simple to add I think, but it might not be worth it...

**Reply by JohnTing** [reply]
For example, Firebasejs's `onValue` method returns an Unsubscribe function that is used to unsubscribe from data updates. (Unsubscribe is a cancel callback)
`function onValue(query: Query, callback: (snapshot: DataSnapshot) => unknown, options: ListenOptions): Unsubscribe;`

This is my use case. I plan to filter data based on options, so I'm recording which methods should be called within x-on:click, and then processing them together using a for loop later on.
But I was initially very puzzled why the filter1 function would execute immediately when I pressed the button, until I realized that Alpine would execute the execution result of the code on its own.
```html
<div x-data="{ filterFuncs: {} }">
        <input type="checkbox" id="c1" x-on:click="filterFuncs[$el.id]=$el.checked ? filter1 : null;">filter by Condition 1</input><br/>
        <input type="checkbox" id="c2" x-on:click="filterFuncs[$el.id]=$el.checked ? filter2 : null;">filter by Condition 2</input><br/>
        <input type="checkbox" id="c3" x-on:click="filterFuncs[$el.id]=$el.checked ? filter3 : null;">filter by Condition 3</input><br/>
</div>
```

**Comment by JohnTing** [comment]
I tested some behavior of Vue3, and it seems that automatic execution only occurs when there is a single function without parentheses.
I don't know how they do this.
in Vue3
```vue
<script setup>
import {ref} from 'vue'

const storedFoo1 = ref(null)
function foo1() {
  console.log('foo1');
  return () => {console.log('foo2');};
}
</script>
<template>
  <button @click="foo1">click</button> <!-- print "foo1" -->
  <button @click="foo1;foo1">click</button> <!-- print nothing -->
  <button @click="foo1()">click</button> <!-- print "foo1" -->
  <button @click="storedFoo1 = foo1">click</button> <!-- print nothing -->
  <button @click="storedFoo1 = foo1()">click</button> <!-- print "foo1" -->
</template>
```

in Alpine
```html
<div x-data="{ storedFoo1: null }">
    <button x-on:click="foo1">click</button> <!-- print "foo1" -->
    <button x-on:click="foo1;foo1">click</button> <!-- print "foo1" -->
    <button x-on:click="foo1()">click</button> <!-- print "foo1" "Foo2" -->
    <button x-on:click="storedFoo1=foo1">click</button><!-- print "foo1" -->
    <button x-on:click="storedFoo1=foo1()">click</button> <!-- print "foo1" "Foo2" -->
</div>

```
